torch_geometric/data/materialized_graph.py

```diff
@@ -0,0 +1,125 @@
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, Optional, Union
+
+from torch import LongTensor
+
+from torch_geometric.typing import EdgeTensorType, EdgeType
+from torch_geometric.utils.mixin import CastMixin
+
+
+class EdgeLayout(Enum):
+    COO = 'COO'
+    CSC = 'CSC'
+    CSR = 'CSR'
+    LIL = 'LIL'
+
+
+@dataclass
+class EdgeAttr(CastMixin):
+    r"""Defines the attributes of an :obj:`MaterializedGraph` edge."""
+
+    # The layout of the edge representation
+    layout: Optional[EdgeLayout] = None
+
+    # The type of the edge
+    edge_type: Optional[Any] = None
+
+
+class MaterializedGraph(MutableMapping):
+    def __init__(self, edge_attr_cls: Any = EdgeAttr):
+        r"""Initializes the materialized graph. Implementor classes can
+        customize the ordering and required nature of their :class:`EdgeAttr`
+        edge attributes by subclassing :class:`EdgeAttr` and passing the
+        subclass as :obj:`attr_cls`."""
+        super().__init__()
+        self.__dict__['_edge_attr_cls'] = edge_attr_cls
+
+    # Core ####################################################################
+
+    @abstractmethod
+    def _put_edge_index(self, edge_index: EdgeTensorType,
+                        edge_attr: EdgeAttr) -> bool:
+        return None
```

Comment:
```suggestion
        raise NotImplementedErrror()
```

Reply:
Yeah thats not needed. A child class has to override methods with `@abstractmethod` decorator.